editor/src/clj/editor/grid.clj

```diff
@@ -13,192 +13,385 @@
+                 u-size (get size-map u-axis-key)
+                 v-size (get size-map v-axis-key)]]
+       (do
+         (gl/gl-color gl (colors/alpha color alpha))
+         (render-grid gl fixed-axis u-size v-size (nth (:aabbs grids) grid-index)))))))
 
 (defn render-scaled-grids
-  [^GL2 gl pass renderables count]
+  [^GL2 gl _pass renderables _count]
   (let [renderable (first renderables)
-        user-render-data (:user-render-data renderable)
-        camera (:camera user-render-data)
-        grids (:grids user-render-data)
+        {:keys [camera grids options]} (:user-render-data renderable)
         view-matrix (c/camera-view-matrix camera)
-        dir         (double-array 4)
-        _           (.getRow view-matrix 2 dir)]
+        dir (double-array 4)
+        _ (.getRow view-matrix 2 dir)]
     (gl/gl-lines gl
-      (render-grid-sizes dir grids)
-      (render-primary-axes (apply geom/aabb-union (:aabbs grids))))))
+      (render-grid-sizes dir grids options)
+      (render-primary-axes (apply geom/aabb-union (:aabbs grids)) options))))
 
-(g/defnk grid-renderable
-  [camera grids]
+(g/defnk produce-renderable
+  [camera grids merged-options]
   {pass/infinity-grid ; Grid lines stretching to infinity. Not depth-clipped to frustum.
    [{:world-transform geom/Identity4d
      :tags #{:grid}
      :render-fn render-scaled-grids
      :user-render-data {:camera camera
-                        :grids grids}}]
+                        :grids grids
+                        :options merged-options}}]
    pass/transparent ; Grid lines potentially intersecting scene geometry.
    [{:world-transform geom/Identity4d
      :tags #{:grid}
      :render-fn render-scaled-grids
      :user-render-data {:camera camera
-                        :grids grids}}]})
-
-(def axis-vectors
-  [(Vector4d. 1.0 0.0 0.0 0.0)
-   (Vector4d. 0.0 1.0 0.0 0.0)
-   (Vector4d. 0.0 0.0 1.0 0.0)])
-
-(defn as-unit-vector
-  [^Vector4d v]
-  (let [v-unit (Vector4d. v)]
-    (set! (. v-unit w) 0)
-    (.normalize v-unit)
-    v-unit))
-
-
-(defn dot
-  ^double [^Vector4d x ^Vector4d y]
-  (.dot x y))
+                        :grids grids
+                        :options merged-options}}]})
 
 (defn frustum-plane-projection
-  [^Vector4d plane1 ^Vector4d plane2]
-  (let [m (Matrix3d. 0.0         0.0         1.0
+  [^Vector4d plane1 ^Vector4d plane2 plane]
+  (let [nx (if (= plane 0) 1.0 0.0)
+        ny (if (= plane 1) 1.0 0.0)
+        nz (if (= plane 2) 1.0 0.0)
+        m (Matrix3d. nx         ny         nz
                      (.x plane1) (.y plane1) (.z plane1)
                      (.x plane2) (.y plane2) (.z plane2))
-        v (Point3d. 0.0 (- (.w plane1)) (- (.w plane2)))]
-    (.invert m)
-    (.transform m v)
-    v))
+        v (Point3d. 0.0 (- (.w plane1)) (- (.w plane2)))
+        det (.determinant m)]
+    (if (< (Math/abs det) 1e-10)
+      (Point3d. 0.0 0.0 0.0)
+      (do (.invert m)
+          (.transform m v)
+          v))))
```

Comment:
While rotating the camera, `invert` would through when the grid was perpendicular to the viewport. The reason seems to be that the matrix cannot be inverted, If the determinant is zero. Also, the editor would sometimes hang if we were getting close to this point. Providing a default when the determinant approaches zero seems to fix both cases.